src/data/nat/fib.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2019 Kevin Kappelmann. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Kappelmann
+-/
+import ..stream.basic
+import .....tactic.norm_num
+import .....tactic.monotonicity
+/-!
+# The Fibonacci Sequence
+
+## Summary
+
+Definition of the Fibonacci sequence `F₀ = 0, F₁ = 1, Fₙ₊₂ = Fₙ + Fₙ₊₁`.
+
+## Main Definitions
+
+- `fib` returns the stream of Fibonacci numbers.
+
+## Main Statements
+
+- `fib.recurrence` shows that `fib` indeed satisfies the Fibonacci recurrence `Fₙ₊₂ = Fₙ + Fₙ₊₁.`.
+
+## Implementation Notes
+
+For efficiency purposes, the sequence is defined using `stream.iterate`.
+
+## Tags
+
+fib, fibonacci
+-/
+
+private def fib.aux_stream : stream (ℕ × ℕ) := stream.iterate (λ p, ⟨p.snd, p.fst + p.snd⟩) ⟨0, 1⟩
+
+/--
+Implementation of the fibonacci sequence satisfying
+`fib 0 = 0, fib 1 = 1, fib (n + 2) = fib n + fib (n + 1)`.
+
+*Note:* We use a stream iterator for better performance when compared to the naive recursive
+implementation.
+-/
+def fib (n : ℕ) : ℕ := (fib.aux_stream n).fst
+
+namespace fib
+
+@[simp]
+lemma zeroth : fib 0 = 0 := rfl
+@[simp]
+lemma first : fib 1 = 1 := rfl
+
+private lemma recurrence_aux {n : ℕ} : (fib.aux_stream (n + 1)).fst = (fib.aux_stream n).snd :=
+begin
+  change (stream.nth (n + 1) $ stream.iterate _ (0, 1)).fst = _,
+  rw [stream.nth_succ_iterate, stream.map_iterate, stream.nth_map, fib.aux_stream]
+end
+
+/-- Shows that `fib` indeed satisfies the Fibonacci recurrence `Fₙ₊₂ = Fₙ + Fₙ₊₁.` -/
+lemma recurrence {n : ℕ} : fib (n + 2) = fib n + fib (n + 1) :=
+begin
+  rw [fib, recurrence_aux],
+  change (stream.nth (n + 1) $ stream.iterate (λ p, _) (0, 1)).snd = _,
+  rw [stream.nth_succ_iterate, stream.map_iterate, stream.nth_map],
+  suffices : (fib.aux_stream n).snd = (stream.iterate (λ p, _) (0, 1) (n + 1)).fst, by
+    simpa only [fib, fib.aux_stream],
+  rw ←recurrence_aux,
+  refl
+end
+
+protected lemma pos {n : ℕ} (n_ne_zero : n ≠ 0) : 0 < fib n :=
+begin
+  induction n with n IH,
+  case nat.zero { contradiction },
+  case nat.succ
+  { cases n,
+    case nat.zero { simp [recurrence, zero_lt_one] },
+    case nat.succ
+    { have : 0 ≤ fib n, by simp,
+      exact (lt_add_of_nonneg_of_lt this $ IH n.succ_ne_zero) }}
+end
+
+lemma nth_fib_le_succ_nth_fib {n : ℕ} : fib n ≤ fib (n + 1) :=
```

Comment:
`fib_le_fib_succ`

Reply:
Probably `fib_mono` would also be useful here.

Reply:
Changed and added.